native/dhseqr.go

```diff
@@ -0,0 +1,255 @@
+// H[unconverged:ihi+1,unconverged:ihi+1] is upper quasi-triangular.
+//
+// If unconverged > 0 and compz == lapack.UpdateZ, then on return
+//  (final Z) = (initial Z) U,
+// where U is the orthogonal matrix in (*) regardless of the value of job.
+//
+// If unconverged > 0 and compz == lapack.InitZ, then on return
+//  (final Z) = U,
+// where U is the orthogonal matrix in (*) regardless of the value of job.
+//
+// References:
+//  [1] R. Byers. LAPACK 3.1 xHSEQR: Tuning and Implementation Notes on the
+//      Small Bulge Multi-Shift QR Algorithm with Aggressive Early Deflation.
+//      LAPACK Working Note 187 (2007)
+//      URL: http://www.netlib.org/lapack/lawnspdf/lawn187.pdf
+//  [2] K. Braman, R. Byers, R. Mathias. The Multishift QR Algorithm. Part I:
+//      Maintaining Well-Focused Shifts and Level 3 Performance. SIAM J. Matrix
+//      Anal. Appl. 23(4) (2002), pp. 929—947
+//      URL: http://dx.doi.org/10.1137/S0895479801384573
+//  [3] K. Braman, R. Byers, R. Mathias. The Multishift QR Algorithm. Part II:
+//      Aggressive Early Deflation. SIAM J. Matrix Anal. Appl. 23(4) (2002), pp. 948—973
+//      URL: http://dx.doi.org/10.1137/S0895479801384585
+//
+// Dhseqr is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dhseqr(job lapack.Job, compz lapack.Comp, n, ilo, ihi int, h []float64, ldh int, wr, wi []float64, z []float64, ldz int, work []float64, lwork int) (unconverged int) {
+	var wantt bool
+	switch job {
+	default:
+		panic(badJob)
+	case lapack.EigenvaluesOnly:
+	case lapack.EigenvaluesAndSchur:
+		wantt = true
+	}
+	var wantz bool
+	switch compz {
+	default:
+		panic("lapack: bad compz")
+	case lapack.None:
+	case lapack.InitZ, lapack.UpdateZ:
+		wantz = true
+	}
+	switch {
+	case n < 0:
+		panic(nLT0)
+	case ilo < 0 || max(0, n-1) < ilo:
+		panic(badIlo)
+	case ihi < min(ilo, n-1) || n <= ihi:
+		panic(badIhi)
+	case len(work) < lwork:
+		panic(shortWork)
+	case lwork < max(1, n) && lwork != -1:
+		panic(badWork)
+	}
+	if lwork != -1 {
+		checkMatrix(n, n, h, ldh)
+		switch {
+		case wantz:
+			checkMatrix(n, n, z, ldz)
+		case len(wr) < n:
+			panic("lapack: wr has insufficient length")
+		case len(wi) < n:
+			panic("lapack: wi has insufficient length")
+		}
+	}
+
+	const (
+		// Matrices of order ntiny or smaller must be processed by
+		// Dlahqr because of insufficient subdiagonal scratch space.
+		// This is a hard limit.
+		ntiny = 11
+
+		// nl is the size of a local workspace to help small matrices
+		// through a rare Dlahqr failure. nl > ntiny is required and
+		// nl <= nmin = Ilaenv(ispec=12,...) is recommended (the default
+		// value of nmin is 75). Using nl = 49 allows up to six
+		// simultaneous shifts and a 16×16 deflation window.
+		nl = 49
+	)
+
+	// Quick return if possible.
+	if n == 0 {
+		work[0] = 1
+		return 0
+	}
+
+	// Quick return in case of a workspace query.
+	impl.Dlaqr04(wantt, wantz, n, ilo, ihi, nil, 0, nil, nil, ilo, ihi, nil, 0, work, -1, 1)
```

Comment:
The Fortran work the way I suggested, which is why I suggested it. It should work because `work` is always mutated by `Dlaqr04` in this function, thus always leaving `work[0]` with the optimum `lwork`.